### PR TITLE
Fix delivering response promises in actor state destructors

### PR DESCRIPTION
**Pull request by dominiklohmann**

### CODE REVIEW

File: libcaf_core/src/response_promise.cpp [hunk 1]
```diff
@@ -166,22 +166,32 @@ void response_promise::state::cancel() {
...
     CAF_LOG_DEBUG("drop response: empty response to asynchronous input");
-  } else if (!stages.empty()) {
+    return;
+  }
+  auto self = weak_self.lock();
+  if (self == nullptr) {
+    auto element = make_mailbox_element(self, id.response_id(),
+                                        std::move(stages),
+                                        std::move(msg));
+    source->enqueue(std::move(element), nullptr);
```
Comment (lava) [inline libcaf_core/src/response_promise.cpp:185]: Don't we have to check `source != nullptr` here as well?

File: libcaf_core/src/response_promise.cpp [hunk 2]
```diff
@@ -166,22 +166,32 @@ void response_promise::state::cancel() {
...
-  } else if (source != nullptr) {
-    detail::profiled_send(self, self->ctrl(), source, id.response_id(),
-                          forwarding_stack{}, self->context(), std::move(msg));
-  }
-  cancel();
+    detail::profiled_send(local_self, std::move(source), next, id, std::move(stages),
+                          local_self->context(), std::move(msg));
+    return;
+  } 
+  CAF_ASSERT(source != nullptr);
```
Comment (lava) [inline libcaf_core/src/response_promise.cpp]: The assumption seems to be that either `source` or `stages` must be non-null, but that doesn't seem to be enforced anywhere by the response promise class? Is it really safe to assert here?

File: libcaf_core/src/response_promise.cpp [hunk 3]
```diff
@@ -166,22 +166,32 @@ void response_promise::state::cancel() {
...
+  }
+  auto self = weak_self.lock();
+  if (self == nullptr) {
+    auto element = make_mailbox_element(self, id.response_id(),
+                                        std::move(stages),
+                                        std::move(msg));
+    source->enqueue(std::move(element), nullptr);
+    return;
+  }
+  auto local_self = static_cast<local_actor*>(weak_self.get()->get());
```
Comment (lava) [inline libcaf_core/src/response_promise.cpp]: `self->get()`?

**Review by lava — APPROVED** [review]
Looks good in principle, although I think I'm a bit confused by the semantics of when `source` can or can't be null.

As an aside, I wonder if this comment from `response_promise.hpp` is also causing races, since the rule doesn't seem to be as clear-cut as the comment suggests:

```
  // Note: response promises must remain local to their owner. Hence, we don't
  //       need a thread-safe reference count for the state.
```

**Comment by dominiklohmann** [comment]
@lava and I went over the changes together in a call to address the review feedback.

**Pull request merged**